File: Exercise_1.java
//Time complexity - O(logN)
//Space Complexity - O(1)

class Solution {
    public int[] searchRange(int[] nums, int target) {

        /** to check first occurence and last occurance of the target */

        int length = nums.length;

        /** check the first Occurence */
        int firstOccurence = this.find(nums,target,true);


        /** return -1 if firstOccurence not found */
        if(firstOccurence == -1) {

            return new int[]{-1,-1};

        }


        /** check the second Occurence */
        int secondOccurence = this.find(nums,target,false);

        /** return firstOccurence & secondOccurence if found */
        return new int[]{firstOccurence,secondOccurence};


    }
    public int find(int[] nums, int target, boolean isFirst) {

        int begin = 0;
        int end = nums.length - 1;


        while(begin <= end){


            int mid = (begin + end) / 2;

            if(nums[mid] == target) {

                if(isFirst) {


                    /** condition to check and only perform binary search on left side */
                    if(mid == begin || nums[mid-1] != target) {

                        return mid;

                    }

                    end = mid - 1;

                }

                else {

                    /** condition to check and only perform binary search on right side */
                    if(mid == end || nums[mid+1] != target) {


                        return mid;
                    }

                    begin = mid + 1;
                }
            }

            else if(nums[mid] > target) {

                end = mid - 1;

            }
            else {
                begin = mid + 1;
            }


        }

        return -1;
    }
}

File: Exercise_1.js
// Time Complexity : O(log n)
// Space Complexity : O(1) 
// Approach : finding element through binaray seaech applying two binary search left to find first and right to find last element

var searchRange = function (nums, target) {
    if (nums == null || nums.length == 0) return [-1, -1]
    const low = 0
    const high = nums.length - 1
    const left = binarySearchLeft(nums, low, high, target)
    if (left == -1) {
        return [-1, -1]
    }
    const right = binarySearchRight(nums, left, high, target)
    return [left, right]
};

function binarySearchLeft(nums, low, high, target) {
    while (low <= high) {
        mid = Math.floor(low + (high - low) / 2)
        if (nums[mid] == target) {
            if (mid == 0 || nums[mid] > nums[mid - 1]) {
                return mid
            } else {
                high = mid - 1;
            }
        } else if (nums[mid] > target) {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    return -1
}

function binarySearchRight(nums, low, high, target) {
    while (low <= high) {
        mid = Math.floor(low + (high - low) / 2)
        if (nums[mid] == target) {
            if (mid == nums.length - 1 || nums[mid] < nums[mid + 1]) {
                return mid
            } else {
                low = mid + 1;
            }
        } else if (nums[mid] > target) {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    return -1
}